### PR TITLE
lambdabot: apply upstream patches for GHC 9.0.2 support

### PR DESCRIPTION
**Pull request by ncfavier**
Tested by building `lambdabot` and running `dice d10` (from the dice plugin) and `yow` (from the fortune plugin) at the prompt.

### CODE REVIEW

**Comment by ncfavier** [comment]
Done. Wasn't sure where to put my changes in `configuration-common.nix` so I picked a random place.

**Comment by maralorn** [comment]
Okay that’s already better.

Now, there is one thing I should have warned you about. We try to not accumulated patches in nixpkgs, but rather use `pkgs.fetchpatch` to pull the patch from github. There are various examples in configuration-common.nix.

Since you are patching different packages in the same commit, if I am not mistaken, this would be a bit finnicky with fetchpatch.
You would need a combination of the `stripLen = 2, extraPrefix = ""` and `includes = [ ... ]` or something like that. And be careful since this is a fixed output derivation you will need to invalidate the hash everytime you modify one of those parameters.

Do you want to try that? If it’s too annoying, I think I could be convinced to merge this as is …

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 1]
```diff
@@ -2372,6 +2372,27 @@ self: super: {
...
+  # compatibility with random-fu 0.3 https://github.com/mokus0/misfortune/pull/5
+  misfortune = appendPatch ./patches/misfortune-ghc9.patch (overrideCabal (drv: {
+    revision = "0";
+    editedCabalFile = "sha256-17uKoAt7bCG3XB2iwi8udOyuci+go/gH1XbxmM3dPoc=";
+  }) super.misfortune);
+
+  # GHC 9 support https://github.com/lambdabot/dice/pull/2
+  dice = appendPatch ./patches/dice-ghc9.patch (overrideCabal (drv: {
+    revision = "0";
+    editedCabalFile = "sha256-Ydw9YiNZ8V0naHjfYhE1m5P7M45wZQu4iTZ5ycf4X4c=";
```
Comment (sternenseemann) [inline pkgs/development/haskell-modules/configuration-common.nix:2386]: To get the original cabal file.

```suggestion
    revision = null;
    editedCabalFile = null;
```

**Comment by ncfavier** [comment]
I managed for dice.

For misfortune the PR is based on master which has a few more commits than the latest hackage release, and the patch doesn't apply cleanly. We would need to fetch two more commits from master for the patch to apply. I also can't just `excludes = [ "misfortune.cabal" ]` because the patch *adds* a dependency to the cabal file.

For lambdabot, what you suggest doesn't seem possible because the `includes` logic happens [*after*](https://github.com/NixOS/nixpkgs/blob/29a988f9781aa746d5afa4a4d3029e22490bb757/pkgs/build-support/fetchpatch/default.nix#L53) the `stripLen` logic for some reason. Unless we make our own `fetchpatch` with `fetchurl` + `runCommand`.

Note that in both cases (misfortune and lambdabot), the only commits on top of the latest release are adjustments to CI, documentation or cabal files, so maybe we could fetch the source from the PR as I did initially and keep the version as-is since there are no code changes.

**Comment by maralorn** [comment]
Yeah, the stripLen/excludes order is really annoying.

Thank you for trying. I guess we should choose our battles wisely and not sink any more work into a working fix.

**Comment by ncfavier** [comment]
> I guess we should choose our battles wisely and not sink any more work into a working fix.

I am not a very wise person https://github.com/NixOS/nixpkgs/pull/165327.

(Let's not block this PR on that one though)

**Comment by maralorn** [comment]
thx for patching

**Pull request merged**